### PR TITLE
Type-Hint http.py

### PR DESCRIPTION
**Pull request by bijij**
## Summary

**This PR is not completed.**

This PR starts the ground-work for type-hinting http.py

In it's current state it relies on ~~#6871~~ and ~~#6872~~ for http payload typings.

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Review by NCPlayz — CHANGES_REQUESTED** [review]
just skimmed over the typings, seems alright

File: discord/http.py [hunk 1]
```diff
@@ -653,35 +737,61 @@ def edit_profile(self, username, avatar):
...
+        reason: Optional[str] = None,
+    ) -> Response[member.Member]:
         r = Route('PATCH', '/guilds/{guild_id}/members/{user_id}', guild_id=guild_id, user_id=user_id)
         payload = {
             'nick': nickname,
         }
         return self.request(r, json=payload, reason=reason)
 
-    def edit_my_voice_state(self, guild_id, payload):
+    def edit_my_voice_state(self, guild_id: Snowflake, payload) -> Response[None]:
```
Comment (AbstractUmbra) [inline discord/http.py:767]: `payload` param should be annotated.

File: discord/http.py [hunk 2]
```diff
@@ -653,35 +737,61 @@ def edit_profile(self, username, avatar):
...
         r = Route('PATCH', '/guilds/{guild_id}/voice-states/@me', guild_id=guild_id)
         return self.request(r, json=payload)
 
-    def edit_voice_state(self, guild_id, user_id, payload):
+    def edit_voice_state(self, guild_id: Snowflake, user_id: Snowflake, payload) -> Response[None]:
```
Comment (AbstractUmbra) [inline discord/http.py:771]: `payload` param should be annotated.

File: discord/http.py [hunk 3]
```diff
@@ -703,11 +813,24 @@ def edit_channel(self, channel_id, *, reason=None, **options):
         payload = {k: v for k, v in options.items() if k in valid_keys}
         return self.request(r, reason=reason, json=payload)
 
-    def bulk_channel_update(self, guild_id, data, *, reason=None):
+    def bulk_channel_update(
+        self,
+        guild_id: Snowflake,
+        data,
```
Comment (AbstractUmbra) [inline discord/http.py:819]: `data` param should be annotated.

File: discord/http.py [hunk 4]
```diff
@@ -772,23 +900,23 @@ def start_private_thread(
...
         return self.request(Route('DELETE', '/channels/{channel_id}/thread-members/@me', channel_id=channel_id))
 
-    def remove_user_from_thread(self, channel_id: int, user_id: int):
+    def remove_user_from_thread(self, channel_id: Snowflake, user_id: Snowflake) -> Response[None]:
         route = Route('DELETE', '/channels/{channel_id}/thread-members/{user_id}', channel_id=channel_id, user_id=user_id)
         return self.request(route)
 
     def get_public_archived_threads(
-        self, channel_id: int, before=None, limit: int = 50
+        self, channel_id: Snowflake, before=None, limit: int = 50
```
Comment (AbstractUmbra) [inline discord/http.py:919]: `before` should perhaps be annotated even if unused?

File: discord/http.py [hunk 5]
```diff
@@ -799,7 +927,7 @@ def get_public_archived_threads(
         return self.request(route, params=params)
 
     def get_private_archived_threads(
-        self, channel_id: int, before=None, limit: int = 50
+        self, channel_id: Snowflake, before=None, limit: int = 50
```
Comment (AbstractUmbra) [inline discord/http.py:930]: `before` should perhaps be annotated even if unused?

File: discord/http.py [hunk 6]
```diff
@@ -810,7 +938,7 @@ def get_private_archived_threads(
         return self.request(route, params=params)
 
     def get_joined_private_archived_threads(
-        self, channel_id: int, before=None, limit: int = 50
+        self, channel_id: Snowflake, before=None, limit: int = 50
```
Comment (AbstractUmbra) [inline discord/http.py:941]: `before` should perhaps be annotated even if unused?

File: discord/http.py [hunk 7]
```diff
@@ -915,19 +1060,19 @@ def edit_guild(self, guild_id, *, reason=None, **fields):
...
-    def get_template(self, code):
+    def get_template(self, code: str) -> Response[template.Template]:
         return self.request(Route('GET', '/guilds/templates/{code}', code=code))
 
-    def guild_templates(self, guild_id):
+    def guild_templates(self, guild_id: Snowflake) -> Response[List[template.Template]]:
         return self.request(Route('GET', '/guilds/{guild_id}/templates', guild_id=guild_id))
 
-    def create_template(self, guild_id, payload):
+    def create_template(self, guild_id: Snowflake, payload) -> Response[template.Template]:
```
Comment (AbstractUmbra) [inline discord/http.py:1069]: `payload` param should be annotated.

File: discord/http.py [hunk 8]
```diff
@@ -1226,11 +1448,15 @@ def get_global_command(self, application_id, command_id) -> Response[interaction
...
-    def upsert_global_command(self, application_id, payload) -> Response[interactions.ApplicationCommand]:
+    def upsert_global_command(self, application_id: Snowflake, payload) -> Response[interactions.ApplicationCommand]:
         r = Route('POST', '/applications/{application_id}/commands', application_id=application_id)
         return self.request(r, json=payload)
 
-    def edit_global_command(self, application_id, command_id, payload) -> Response[interactions.ApplicationCommand]:
+    def edit_global_command(self,
+        application_id: Snowflake,
+        command_id: Snowflake,
+        payload,
```
Comment (AbstractUmbra) [inline discord/http.py:1458]: `payload` param should be annotated?

File: discord/http.py [hunk 9]
```diff
@@ -1279,7 +1510,12 @@ def get_guild_command(self, application_id, guild_id, command_id) -> Response[in
         )
         return self.request(r)
 
-    def upsert_guild_command(self, application_id, guild_id, payload) -> Response[interactions.ApplicationCommand]:
+    def upsert_guild_command(
+        self,
+        application_id: Snowflake,
+        guild_id: Snowflake,
+        payload,
```
Comment (AbstractUmbra) [inline discord/http.py:1517]: `payload` param should be annotated?

File: discord/http.py [hunk 10]
```diff
@@ -1288,7 +1524,12 @@ def upsert_guild_command(self, application_id, guild_id, payload) -> Response[in
         )
         return self.request(r, json=payload)
 
-    def edit_guild_command(self, application_id, guild_id, command_id, payload) -> Response[interactions.ApplicationCommand]:
+    def edit_guild_command(self,
+        application_id: Snowflake,
+        guild_id: Snowflake,
+        command_id: Snowflake,
+        payload,
```
Comment (AbstractUmbra) [inline discord/http.py:1531]: `payload` param should be annotated?

File: discord/http.py [hunk 11]
```diff
@@ -1315,7 +1561,10 @@ def delete_guild_command(self, application_id, guild_id, command_id):
         return self.request(r)
 
     def bulk_upsert_guild_commands(
-        self, application_id, guild_id, payload
+        self, 
+        application_id: Snowflake,
+        guild_id: Snowflake, 
+        payload,
```
Comment (AbstractUmbra) [inline discord/http.py:1567]: `payload` param should be annotated.

File: discord/http.py [hunk 12]
```diff
@@ -1495,7 +1756,13 @@ def get_application_command_permissions(
         )
         return self.request(r)
 
-    def edit_application_command_permissions(self, application_id, guild_id, command_id, payload):
+    def edit_application_command_permissions(
+        self,
+        application_id: Snowflake,
+        guild_id: Snowflake,
+        command_id: Snowflake,
+        payload,
```
Comment (AbstractUmbra) [inline discord/http.py:1764]: `payload` param should be annotated.

File: discord/http.py [hunk 13]
```diff
@@ -1505,7 +1772,12 @@ def edit_application_command_permissions(self, application_id, guild_id, command
         )
         return self.request(r, json=payload)
 
-    def bulk_edit_guild_application_command_permissions(self, application_id, guild_id, payload):
+    def bulk_edit_guild_application_command_permissions(
+        self,
+        application_id: Snowflake,
+        guild_id: Snowflake,
+        payload
```
Comment (AbstractUmbra) [inline discord/http.py:1779]: `payload` param should be annotated.

**Review by AbstractUmbra — COMMENTED** [review]
A few annotation skips.

File: discord/http.py [hunk 14]
```diff
@@ -471,23 +528,23 @@ def send_files(
             components=components,
         )
 
-    def delete_message(self, channel_id, message_id, *, reason=None):
+    def delete_message(self, channel_id: Snowflake, message_id: Snowflake, *, reason: Optional[str] = None) -> Response[None]:
         r = Route('DELETE', '/channels/{channel_id}/messages/{message_id}', channel_id=channel_id, message_id=message_id)
         return self.request(r, reason=reason)
 
-    def delete_messages(self, channel_id, message_ids, *, reason=None):
+    def delete_messages(self, channel_id: Snowflake, message_ids: List[int], *, reason: Optional[str] = None) -> Response[None]:
```
Comment (Rapptz) [inline discord/http.py:535]:
```suggestion
    def delete_messages(self, channel_id: Snowflake, message_ids: List[Snowflake], *, reason: Optional[str] = None) -> Response[None]:
```

File: discord/http.py [hunk 15]
```diff
@@ -987,22 +1141,35 @@ def prune_members(self, guild_id, days, compute_prune_count, roles, *, reason=No
...
         return self.request(Route('GET', '/guilds/{guild_id}/emojis/{emoji_id}', guild_id=guild_id, emoji_id=emoji_id))
 
-    def create_custom_emoji(self, guild_id, name, image, *, roles=None, reason=None):
+    def create_custom_emoji(
+        self,
+        guild_id: Snowflake,
+        name: str,
+        image: bytes,
+        *,
+        roles: Optional[List[int]] = None,
```
Comment (Rapptz) [inline discord/http.py:1170]:
```suggestion
        roles: Optional[List[Snowflake]] = None,
```

File: discord/http.py [hunk 16]
```diff
@@ -1012,20 +1179,33 @@ def create_custom_emoji(self, guild_id, name, image, *, roles=None, reason=None)
...
         r = Route('DELETE', '/guilds/{guild_id}/emojis/{emoji_id}', guild_id=guild_id, emoji_id=emoji_id)
         return self.request(r, reason=reason)
 
-    def edit_custom_emoji(self, guild_id: int, emoji_id: int, *, payload, reason: Optional[str] = None):
+    def edit_custom_emoji(
+        self,
+        guild_id: Snowflake,
+        emoji_id: Snowflake,
+        *,
+        payload,
```
Comment (Rapptz) [inline discord/http.py:1197]:
```suggestion
        payload: Dict[str, Any],
```

File: discord/http.py [hunk 17]
```diff
@@ -1122,36 +1310,49 @@ def invites_from(self, guild_id: Snowflake) -> Response[List[invite.Invite]]:
...
-    def create_role(self, guild_id, *, reason=None, **fields):
+    def create_role(self, guild_id: Snowflake, *, reason: Optional[str] = None, **fields: Any) -> Response[role.Role]:
         r = Route('POST', '/guilds/{guild_id}/roles', guild_id=guild_id)
         return self.request(r, json=fields, reason=reason)
 
-    def move_role_position(self, guild_id, positions, *, reason=None):
+    def move_role_position(
+        self,
+        guild_id: Snowflake,
+        positions,
```
Comment (Rapptz) [inline discord/http.py:1348]:
```suggestion
        positions: List[Dict[str, Any]],
```

File: discord/http.py [hunk 18]
```diff
@@ -1122,36 +1309,49 @@ def invites_from(self, guild_id: Snowflake) -> Response[List[invite.Invite]]:
...
+    def delete_role(self, guild_id: Snowflake, role_id: Snowflake, *, reason: Optional[str] = None) -> Response[None]:
         r = Route('DELETE', '/guilds/{guild_id}/roles/{role_id}', guild_id=guild_id, role_id=role_id)
         return self.request(r, reason=reason)
 
-    def replace_roles(self, user_id, guild_id, role_ids, *, reason=None):
+    def replace_roles(
+        self,
+        user_id: Snowflake,
+        guild_id: Snowflake,
+        role_ids: List[int],
```
[the remark this answers is not in the document]
Comment (Rapptz) [reply discord/http.py:1335]:
```suggestion
        role_ids: List[Snowflake],
```

**Review by Rapptz — CHANGES_REQUESTED** [review]

File: discord/http.py [hunk 19]
```diff
@@ -67,79 +106,90 @@ async def json_or_text(response):
...
 
-    def defer(self):
+    def defer(self) -> None:
         self._unlock = False
 
-    def __exit__(self, type, value, traceback):
+    def __exit__(self,
+        exc_type: Optional[Type[BE]],
+        exc: Optional[BE],
+        traceback: TracebackType,
```
Comment (ToxicKidz) [inline discord/http.py:145]:
```suggestion
        traceback: Optional[TracebackType],
```

File: discord/http.py [hunk 20]
```diff
@@ -154,7 +204,14 @@ async def ws_connect(self, url, *, compress=0):
 
         return await self.__session.ws_connect(url, **kwargs)
 
-    async def request(self, route, *, files=None, form=None, **kwargs) -> Any:
+    async def request(
+        self,
+        route: Route,
+        *,
+        files: Optional[Sequence[File]] = None,
+        form: Iterable[Dict[str, Any]] = None,
```
Comment (ToxicKidz) [inline discord/http.py:212]:
```suggestion
        form: Optional[Iterable[Dict[str, Any]]] = None,
```

File: discord/http.py [hunk 21]
```diff
@@ -447,17 +504,17 @@ def send_multipart_helper(
...
-        message_reference=None,
-        components=None,
-    ):
+        files: Sequence[File],
+        content: Optional[str] = None,
+        tts: bool = False,
+        embed: Optional[embed.Embed] = None,
+        nonce: Optional[str] = None,
+        allowed_mentions: Optional[message.AllowedMentions] = None,
+        message_reference: Optional[Snowflake] = None,
```
Comment (ToxicKidz) [inline discord/http.py:515]: Isn't `message_reference` a dict?
```suggestion
        message_reference: Optional[message.MessageReference] = None,
```

File: discord/http.py [hunk 22]
```diff
@@ -518,7 +575,14 @@ def remove_own_reaction(self, channel_id, message_id, emoji):
...
         return self.request(r)
 
-    def get_reaction_users(self, channel_id, message_id, emoji, limit, after=None):
+    def get_reaction_users(
+        self,
+        channel_id: Snowflake,
+        message_id: Snowflake,
+        emoji: str,
+        limit: int,
+        after=None,
```
Comment (ToxicKidz) [inline discord/http.py:584]:
```suggestion
        after: Optional[Snowflake] = None,
```

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

**Review by Rapptz — APPROVED** [review]

**Pull request merged**